google/cloud/internal/url_encode.cc

```diff
@@ -24,35 +23,54 @@ namespace internal {
 
 namespace {
 
-using UrlMappings =
-    std::array<std::pair<absl::string_view, absl::string_view>, 25>;
-
-constexpr auto kEncodeMappings = UrlMappings{{
-    {" ", "%20"}, {"\"", "%22"}, {"#", "%23"},  {"$", "%24"}, {"%", "%25"},
-    {"&", "%26"}, {"+", "%2B"},  {",", "%2C"},  {"/", "%2F"}, {":", "%3A"},
-    {";", "%3B"}, {"<", "%3C"},  {"=", "%3D"},  {">", "%3E"}, {"?", "%3F"},
-    {"@", "%40"}, {"[", "%5B"},  {"\\", "%5C"}, {"]", "%5D"}, {"^", "%5E"},
-    {"`", "%60"}, {"{", "%7B"},  {"|", "%7C"},  {"}", "%7D"}, {"\177", "%7F"},
-}};
-
-// Until we require C++17, it isn't worth trying to eliminate the duplication
-// using a constexpr function to produce kDecodeMappings from kEncodeMappings.
-constexpr auto kDecodeMappings = UrlMappings{{
-    {"%20", " "}, {"%22", "\""}, {"%23", "#"},  {"%24", "$"}, {"%25", "%"},
-    {"%26", "&"}, {"%2B", "+"},  {"%2C", ","},  {"%2F", "/"}, {"%3A", ":"},
-    {"%3B", ";"}, {"%3C", "<"},  {"%3D", "="},  {"%3E", ">"}, {"%3F", "?"},
-    {"%40", "@"}, {"%5B", "["},  {"%5C", "\\"}, {"%5D", "]"}, {"%5E", "^"},
-    {"%60", "`"}, {"%7B", "{"},  {"%7C", "|"},  {"%7D", "}"}, {"%7F", "\177"},
-}};
+bool WeNeedToEscape(unsigned char c) {
+  static auto const* const kSet = new std::set<char>{
+      ' ', '\"', '#', '$', '%', '&',  '+', ',', '/', ':', ';', '<',
+      '=', '>',  '?', '@', '[', '\\', ']', '^', '`', '{', '|', '}',
+  };
```

Comment:
You may be falling for the same _"I think I was trying to use the standard library as much as possible"_ trap here.  For example, if you just `switch (c)` and list those cases, you can utilize the compiler's smarts to pick the best of linear search, binary search, or table lookup for that set of values.  My rough measurements indicate that would be at least twice as fast as the `std::set` lookup.

Reply:
Done.